From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: nicollasgtb <nicolas2006.capela@gmail.com>
Date: Sat, 13 Dec 2025 00:56:22 -0300
Subject: [PATCH] Fix East West Cannoning


diff --git a/src/main/java/com/hpfxd/pandaspigot/config/PandaSpigotWorldConfig.java b/src/main/java/com/hpfxd/pandaspigot/config/PandaSpigotWorldConfig.java
index 360ed37b95fce68941c234615b465090b8ea7308..9e4e981f71f04e3badfe87663a496bc77cf6afd3 100644
--- a/src/main/java/com/hpfxd/pandaspigot/config/PandaSpigotWorldConfig.java
+++ b/src/main/java/com/hpfxd/pandaspigot/config/PandaSpigotWorldConfig.java
@@ -58,6 +58,8 @@ public class PandaSpigotWorldConfig {
     public boolean optimizeLiquidExplosions = true; // This seems like a pretty sane default
 
     public boolean optimizeArmorStandMovement = false; // Doesn't fully emulate vanilla behavior, see TacoSpigot issue #1
+    
+    public boolean fixEastWestCannoning = false;
 	
     @Comment("This option controls whether or not there is a chance for arrow crits to deal extra damage.\n" +
             "By default, this is true (vanilla behaviour)")
diff --git a/src/main/java/net/minecraft/server/Entity.java b/src/main/java/net/minecraft/server/Entity.java
index 9d8bb72b469163db7def78f623d18dc4056b1f6b..b2258bcb7599ed47e74fe76b123d01cab2867846 100644
--- a/src/main/java/net/minecraft/server/Entity.java
+++ b/src/main/java/net/minecraft/server/Entity.java
@@ -446,31 +446,15 @@ public abstract class Entity implements ICommandListener {
 
 
     public void move(double d0, double d1, double d2) {
-        if (this.loadChunks) loadChunks(); // PaperSpigot - Load chunks
+        if (this.loadChunks) {
+            loadChunks(); // PaperSpigot - Load chunks
+        }
+
         if (this.noclip) {
             this.a(this.getBoundingBox().c(d0, d1, d2));
             this.recalcPosition();
         } else {
-            // PandaSpigot start - Fix collision detection of entities with blocks, moved back to below
-            /*
-            // CraftBukkit start - Don't do anything if we aren't moving
-            // We need to do this regardless of whether or not we are moving thanks to portals
-            try {
-                this.checkBlockCollisions();
-            } catch (Throwable throwable) {
-                CrashReport crashreport = CrashReport.a(throwable, "Checking entity block collision");
-                CrashReportSystemDetails crashreportsystemdetails = crashreport.a("Entity being checked for collision");
-
-                this.appendEntityCrashDetails(crashreportsystemdetails);
-                throw new ReportedException(crashreport);
-            }
-            // Check if we're moving
-            if (d0 == 0 && d1 == 0 && d2 == 0 && this.vehicle == null && this.passenger == null) {
-                return;
-            }
-            */
-            // PandaSpigot end
-            // CraftBukkit end
+            // PandaSpigot start - Fix collision detection of entities with blocks
             this.world.methodProfiler.a("move");
             double d3 = this.locX;
             double d4 = this.locY;
@@ -524,6 +508,7 @@ public abstract class Entity implements ICommandListener {
                     }
 
                     d6 = d0;
+
                     if (d2 < d9 && d2 >= -d9) {
                         d2 = 0.0D;
                     } else if (d2 > 0.0D) {
@@ -540,12 +525,10 @@ public abstract class Entity implements ICommandListener {
             double yLength = totalArea.e - totalArea.b;
             double zLength = totalArea.f - totalArea.c;
             boolean axisScan = this.world.pandaSpigotConfig.optimizeTntMovement && xLength * yLength * zLength > 10;
-
             List list = this.world.getCubes(this, axisScan ? this.getBoundingBox().a(0, d1, 0) : totalArea);
             // PandaSpigot end
 
             AxisAlignedBB axisalignedbb = this.getBoundingBox();
-
             AxisAlignedBB axisalignedbb1;
 
             for (Iterator iterator = list.iterator(); iterator.hasNext(); d1 = axisalignedbb1.b(this.getBoundingBox(), d1)) {
@@ -554,25 +537,48 @@ public abstract class Entity implements ICommandListener {
 
             this.a(this.getBoundingBox().c(0.0D, d1, 0.0D));
             boolean flag1 = this.onGround || d7 != d1 && d7 < 0.0D;
-
-            if (axisScan) list = this.world.getCubes(this, this.getBoundingBox().a(d0, 0, 0)); // PandaSpigot - get x axis blocks
-
             AxisAlignedBB axisalignedbb2;
             Iterator iterator1;
 
-            for (iterator1 = list.iterator(); iterator1.hasNext(); d0 = axisalignedbb2.a(this.getBoundingBox(), d0)) {
-                axisalignedbb2 = (AxisAlignedBB) iterator1.next();
-            }
+            // TacoSpigot Logic Applied Here (Controlled by Config)
+            if (this.world.pandaSpigotConfig.fixEastWestCannoning && Math.abs(d0) > Math.abs(d2)) { 
+                // Process Z Axis First (Fixes corner alignment)
+                if (axisScan) {
+                    list = this.world.getCubes(this, this.getBoundingBox().a(0, 0, d2));
+                }
+                for (iterator1 = list.iterator(); iterator1.hasNext(); d2 = axisalignedbb2.c(this.getBoundingBox(), d2)) {
+                    axisalignedbb2 = (AxisAlignedBB) iterator1.next();
+                }
+                this.a(this.getBoundingBox().c(0.0D, 0.0D, d2));
 
-            this.a(this.getBoundingBox().c(d0, 0.0D, 0.0D));
+                // Process X Axis Second
+                if (axisScan) {
+                    list = this.world.getCubes(this, this.getBoundingBox().a(d0, 0, 0));
+                }
+                for (iterator1 = list.iterator(); iterator1.hasNext(); d0 = axisalignedbb2.a(this.getBoundingBox(), d0)) {
+                    axisalignedbb2 = (AxisAlignedBB) iterator1.next();
+                }
+                this.a(this.getBoundingBox().c(d0, 0.0D, 0.0D));
 
-            if (axisScan) list = this.world.getCubes(this, this.getBoundingBox().a(0, 0, d2)); // PandaSpigot - get z axis blocks
+            } else {
+                // Original Logic / Vanilla (X first, then Z)
+                if (axisScan) {
+                    list = this.world.getCubes(this, this.getBoundingBox().a(d0, 0, 0));
+                }
+                for (iterator1 = list.iterator(); iterator1.hasNext(); d0 = axisalignedbb2.a(this.getBoundingBox(), d0)) {
+                    axisalignedbb2 = (AxisAlignedBB) iterator1.next();
+                }
+                this.a(this.getBoundingBox().c(d0, 0.0D, 0.0D));
 
-            for (iterator1 = list.iterator(); iterator1.hasNext(); d2 = axisalignedbb2.c(this.getBoundingBox(), d2)) {
-                axisalignedbb2 = (AxisAlignedBB) iterator1.next();
+                if (axisScan) {
+                    list = this.world.getCubes(this, this.getBoundingBox().a(0, 0, d2));
+                }
+                for (iterator1 = list.iterator(); iterator1.hasNext(); d2 = axisalignedbb2.c(this.getBoundingBox(), d2)) {
+                    axisalignedbb2 = (AxisAlignedBB) iterator1.next();
+                }
+                this.a(this.getBoundingBox().c(0.0D, 0.0D, d2));
             }
 
-            this.a(this.getBoundingBox().c(0.0D, 0.0D, d2));
             if (this.S > 0.0F && flag1 && (d6 != d0 || d8 != d2)) {
                 double d10 = d0;
                 double d11 = d1;
@@ -585,7 +591,6 @@ public abstract class Entity implements ICommandListener {
                 AxisAlignedBB axisalignedbb4 = this.getBoundingBox();
                 AxisAlignedBB axisalignedbb5 = axisalignedbb4.a(d6, 0.0D, d8);
                 double d13 = d1;
-
                 AxisAlignedBB axisalignedbb6;
 
                 for (Iterator iterator2 = list1.iterator(); iterator2.hasNext(); d13 = axisalignedbb6.b(axisalignedbb5, d13)) {
@@ -594,7 +599,6 @@ public abstract class Entity implements ICommandListener {
 
                 axisalignedbb4 = axisalignedbb4.c(0.0D, d13, 0.0D);
                 double d14 = d6;
-
                 AxisAlignedBB axisalignedbb7;
 
                 for (Iterator iterator3 = list1.iterator(); iterator3.hasNext(); d14 = axisalignedbb7.a(axisalignedbb4, d14)) {
@@ -603,7 +607,6 @@ public abstract class Entity implements ICommandListener {
 
                 axisalignedbb4 = axisalignedbb4.c(d14, 0.0D, 0.0D);
                 double d15 = d8;
-
                 AxisAlignedBB axisalignedbb8;
 
                 for (Iterator iterator4 = list1.iterator(); iterator4.hasNext(); d15 = axisalignedbb8.c(axisalignedbb4, d15)) {
@@ -613,7 +616,6 @@ public abstract class Entity implements ICommandListener {
                 axisalignedbb4 = axisalignedbb4.c(0.0D, 0.0D, d15);
                 AxisAlignedBB axisalignedbb9 = this.getBoundingBox();
                 double d16 = d1;
-
                 AxisAlignedBB axisalignedbb10;
 
                 for (Iterator iterator5 = list1.iterator(); iterator5.hasNext(); d16 = axisalignedbb10.b(axisalignedbb9, d16)) {
@@ -622,7 +624,6 @@ public abstract class Entity implements ICommandListener {
 
                 axisalignedbb9 = axisalignedbb9.c(0.0D, d16, 0.0D);
                 double d17 = d6;
-
                 AxisAlignedBB axisalignedbb11;
 
                 for (Iterator iterator6 = list1.iterator(); iterator6.hasNext(); d17 = axisalignedbb11.a(axisalignedbb9, d17)) {
@@ -631,7 +632,6 @@ public abstract class Entity implements ICommandListener {
 
                 axisalignedbb9 = axisalignedbb9.c(d17, 0.0D, 0.0D);
                 double d18 = d8;
-
                 AxisAlignedBB axisalignedbb12;
 
                 for (Iterator iterator7 = list1.iterator(); iterator7.hasNext(); d18 = axisalignedbb12.c(axisalignedbb9, d18)) {
@@ -661,6 +661,7 @@ public abstract class Entity implements ICommandListener {
                 }
 
                 this.a(this.getBoundingBox().c(0.0D, d1, 0.0D));
+
                 if (d10 * d10 + d12 * d12 >= d0 * d0 + d2 * d2) {
                     d0 = d10;
                     d1 = d11;
@@ -684,7 +685,6 @@ public abstract class Entity implements ICommandListener {
 
             if (block.getMaterial() == Material.AIR) {
                 Block block1 = this.world.getType(blockposition.down()).getBlock();
-
                 if (block1 instanceof BlockFence || block1 instanceof BlockCobbleWall || block1 instanceof BlockFenceGate) {
                     block = block1;
                     blockposition = blockposition.down();
@@ -692,6 +692,7 @@ public abstract class Entity implements ICommandListener {
             }
 
             this.a(d1, this.onGround, block, blockposition);
+
             if (d6 != d0) {
                 this.motX = 0.0D;
             }
@@ -739,8 +740,10 @@ public abstract class Entity implements ICommandListener {
 
                 this.M = (float) ((double) this.M + (double) MathHelper.sqrt(d21 * d21 + d23 * d23) * 0.6D);
                 this.N = (float) ((double) this.N + (double) MathHelper.sqrt(d21 * d21 + d22 * d22 + d23 * d23) * 0.6D);
+
                 if (this.N > (float) this.h && block.getMaterial() != Material.AIR) {
                     this.h = (int) this.N + 1;
+
                     if (this.V()) {
                         float f = MathHelper.sqrt(this.motX * this.motX * 0.20000000298023224D + this.motY * this.motY + this.motZ * this.motZ * 0.20000000298023224D) * 0.35F;
 
@@ -763,7 +766,6 @@ public abstract class Entity implements ICommandListener {
             } catch (Throwable throwable) {
                 CrashReport crashreport = CrashReport.a(throwable, "Checking entity block collision");
                 CrashReportSystemDetails crashreportsystemdetails = crashreport.a("Entity being checked for collision");
-
                 this.appendEntityCrashDetails(crashreportsystemdetails);
                 throw new ReportedException(crashreport);
             }
@@ -774,6 +776,7 @@ public abstract class Entity implements ICommandListener {
 
             if (this.world.e(this.getBoundingBox().shrink(0.001D, 0.001D, 0.001D))) {
                 this.burn(1);
+
                 if (!flag2) {
                     ++this.fireTicks;
                     // CraftBukkit start - Not on fire yet
