From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: nicollasgtb <nicolas2006.capela@gmail.com>
Date: Wed, 17 Dec 2025 15:49:05 -0300
Subject: [PATCH] Fix mob spawner delay reset


diff --git a/src/main/java/net/minecraft/server/MobSpawnerAbstract.java b/src/main/java/net/minecraft/server/MobSpawnerAbstract.java
index c74bd1b944b6628d315a6f5b7b745ccc4133a6be..bd5b2be196651c8a63f007fd7a1b6cf1d2f6c2e6 100644
--- a/src/main/java/net/minecraft/server/MobSpawnerAbstract.java
+++ b/src/main/java/net/minecraft/server/MobSpawnerAbstract.java
@@ -79,7 +79,7 @@ public abstract class MobSpawnerAbstract {
                 this.f = this.e;
                 this.e = (this.e + (double) (1000.0F / ((float) this.spawnDelay + 200.0F))) % 360.0D;
             } else {
-                if (this.spawnDelay < -tickDelay) { // PaperSpigot
+                if (this.spawnDelay <= -1) { // PaperSpigot
                     this.h();
                 }
 
