From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: theFoxley <7950977+theFoxley@users.noreply.github.com>
Date: Sat, 10 Jan 2026 11:00:24 -0300
Subject: [PATCH] Prevent desync with some packets that use entity id and check
 for 0


diff --git a/src/main/java/net/minecraft/server/Entity.java b/src/main/java/net/minecraft/server/Entity.java
index b2258bcb7599ed47e74fe76b123d01cab2867846..5ff3e1086f90ca7ba067ea6bf982ee903c7b8f49 100644
--- a/src/main/java/net/minecraft/server/Entity.java
+++ b/src/main/java/net/minecraft/server/Entity.java
@@ -61,7 +61,8 @@ public abstract class Entity implements ICommandListener {
     // PandaSpigot end
 
     private static final AxisAlignedBB a = new AxisAlignedBB(0.0D, 0.0D, 0.0D, 0.0D, 0.0D, 0.0D);
-    private static int entityCount;
+    // SportPaper - set starting id to 1 to prevent desync in some packets (PacketPlayOutSpawnEntity...)
+    public static int entityCount = 1; // SportPaper - make public
     private int id;
     public double j;
     public boolean k;
