From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Heath Logan Campbell <github@heathlogancampbell.com>
Date: Sat, 10 Jan 2026 11:02:05 -0300
Subject: [PATCH] Optimize networking by using IP_TOS


diff --git a/src/main/java/net/minecraft/server/ServerConnection.java b/src/main/java/net/minecraft/server/ServerConnection.java
index 882bf62cc8a08e51b5d315c608194a68ae025489..45f1139d70875ec71166b72cef8ab2ad51eb7e21 100644
--- a/src/main/java/net/minecraft/server/ServerConnection.java
+++ b/src/main/java/net/minecraft/server/ServerConnection.java
@@ -110,6 +110,7 @@ public class ServerConnection {
                 protected void initChannel(Channel channel) throws Exception {
                     try {
                         channel.config().setOption(ChannelOption.TCP_NODELAY, Boolean.valueOf(true));
+                        channel.config().setOption(ChannelOption.IP_TOS, 0x18); // SportPaper - Optimize networking by using IP_TOS
                     } catch (ChannelException channelexception) {
                         ;
                     }
