From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: codenicollas <work.nicollas@gmail.com>
Date: Wed, 25 Feb 2026 18:07:34 -0300
Subject: [PATCH] Optimize AI ChunkCache by Lithium


diff --git a/src/main/java/net/minecraft/server/ChunkCache.java b/src/main/java/net/minecraft/server/ChunkCache.java
index 97acfc07ead11e4f692603ee1dca4465b1c44427..809ed09d5046f160a932a7f50727fd86c556ee47 100644
--- a/src/main/java/net/minecraft/server/ChunkCache.java
+++ b/src/main/java/net/minecraft/server/ChunkCache.java
@@ -2,6 +2,13 @@ package net.minecraft.server;
 
 public class ChunkCache implements IBlockAccess {
 
+    // PandaSpigot start - Lithium: Optimize AI ChunkCache
+    private static final IBlockData DEFAULT_BLOCK = Blocks.AIR.getBlockData();
+    private Chunk[] chunksFlat;
+    private int xLen;
+    private int zLen;
+    // PandaSpigot end
+
     protected int a;
     protected int b;
     protected Chunk[][] c;
@@ -34,6 +41,15 @@ public class ChunkCache implements IBlockAccess {
             }
         }
 
+        // PandaSpigot start - Lithium: Optimize AI ChunkCache
+        this.xLen = this.c.length;
+        this.zLen = this.c[0].length;
+        this.chunksFlat = new Chunk[this.xLen * this.zLen];
+
+        for (int x = 0; x < this.xLen; x++) {
+            System.arraycopy(this.c[x], 0, this.chunksFlat, x * this.zLen, this.zLen);
+        }
+        // PandaSpigot end
     }
 
     @Override
@@ -46,20 +62,31 @@ public class ChunkCache implements IBlockAccess {
 
     @Override
     public IBlockData getType(BlockPosition blockposition) {
-        if (blockposition.getY() >= 0 && blockposition.getY() < 256) {
-            int i = (blockposition.getX() >> 4) - this.a;
-            int j = (blockposition.getZ() >> 4) - this.b;
+        // PandaSpigot start - Lithium: Optimize AI ChunkCache
+        int y = blockposition.getY();
 
-            if (i >= 0 && i < this.c.length && j >= 0 && j < this.c[i].length) {
-                Chunk chunk = this.c[i][j];
+        if (y >= 0 && y < 256) {
+            int x = blockposition.getX();
+            int z = blockposition.getZ();
+
+            int chunkX = (x >> 4) - this.a;
+            int chunkZ = (z >> 4) - this.b;
+
+            if (chunkX >= 0 && chunkX < this.xLen && chunkZ >= 0 && chunkZ < this.zLen) {
+                Chunk chunk = this.chunksFlat[(chunkX * this.zLen) + chunkZ];
 
                 if (chunk != null) {
-                    return chunk.getBlockData(blockposition);
+                    ChunkSection section = chunk.getSections()[y >> 4];
+
+                    if (section != null) {
+                        return section.getType(x & 15, y & 15, z & 15);
+                    }
                 }
             }
         }
 
-        return Blocks.AIR.getBlockData();
+        return DEFAULT_BLOCK;
+        // PandaSpigot end
     }
 
     @Override
