From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: codenicollas <work.nicollas@gmail.com>
Date: Sat, 21 Feb 2026 15:42:13 -0300
Subject: [PATCH] Adaptative chunk GC

diff --git a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
index b1e389f16a2d5a85a15bb934117f4ba207c856b9..247736f80497382d7cf27a5c09d62f995b22bb0d 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
@@ -1418,7 +1418,14 @@ public class CraftWorld implements World {
     public void processChunkGC() {
         chunkGCTickCount++;
 
-        if (chunkLoadCount >= server.chunkGCLoadThresh && server.chunkGCLoadThresh > 0) {
+        // PandaSpigot start - Adaptative chunk GC
+        int playerCount = getPlayers().size();
+        int viewDistance = getHandle().getServer().getViewDistance();
+        int viewChunks = ((viewDistance * 2) + 1);
+        int chunkGCLoadThreshold = org.github.paperspigot.PaperSpigotConfig.adaptativeChunkGC ? (world.keepSpawnInMemory ? 256 : 0) + playerCount * (viewChunks * viewChunks) : server.chunkGCLoadThresh;
+
+        if (chunkLoadCount > chunkGCLoadThreshold && chunkGCLoadThreshold > 0) {
+        // PandaSpigot end
             chunkLoadCount = 0;
         } else if (chunkGCTickCount >= server.chunkGCPeriod && server.chunkGCPeriod > 0) {
             chunkGCTickCount = 0;
diff --git a/src/main/java/org/github/paperspigot/PaperSpigotConfig.java b/src/main/java/org/github/paperspigot/PaperSpigotConfig.java
index 7962e70e26b43c21d3fad0af74b100d9253741d5..a8abe9d2751e0faaad29e75ebd508c82675a2cda 100644
--- a/src/main/java/org/github/paperspigot/PaperSpigotConfig.java
+++ b/src/main/java/org/github/paperspigot/PaperSpigotConfig.java
@@ -174,6 +174,11 @@ public class PaperSpigotConfig
         regionFileCacheSize = getInt("settings.region-file-cache-size", 256);
     }
 
+    public static boolean adaptativeChunkGC;
+    private static void adaptativeChunkGC() {
+        adaptativeChunkGC = getBoolean("adaptative-chunk-gc", true);
+    }
+
     public static double strengthEffectModifier;
     public static double weaknessEffectModifier;
     private static void effectModifiers()
